specerror/error.go

```diff
@@ -55,56 +122,140 @@ const (
+		return fmt.Sprintf(referenceTemplate, version, "config.md#poststart"), nil
+	}
+	poststopRef = func(version string) (reference string, err error) {
+		return fmt.Sprintf(referenceTemplate, version, "config.md#poststop"), nil
+	}
+	annotationsRef = func(version string) (reference string, err error) {
+		return fmt.Sprintf(referenceTemplate, version, "config.md#annotations"), nil
+	}
+	extensibilityRef = func(version string) (reference string, err error) {
+		return fmt.Sprintf(referenceTemplate, version, "config.md#extensibility"), nil
+	}
+	validValuesRef = func(version string) (reference string, err error) {
+		return fmt.Sprintf(referenceTemplate, version, "config.md#valid-values"), nil
+	}
+
 	defaultFSRef = func(version string) (reference string, err error) {
 		return fmt.Sprintf(referenceTemplate, version, "config-linux.md#default-filesystems"), nil
 	}
+
 	runtimeCreateRef = func(version string) (reference string, err error) {
 		return fmt.Sprintf(referenceTemplate, version, "runtime.md#create"), nil
 	}
 )
 
+type errorTemplate struct {
+	Level     rfc2119.Level
+	Reference func(version string) (reference string, err error)
+}
+
+// Error represents a runtime-spec violation.
+type Error struct {
+	// Err holds the RFC 2119 violation.
+	Err rfc2119.Error
+
+	// Code is a matchable holds a Code
+	Code Code
+}
+
 var ociErrors = map[Code]errorTemplate{
 	// Bundle.md
 	// Container Format
 	ConfigFileExistence:  {Level: rfc2119.Must, Reference: containerFormatRef},
 	ArtifactsInSingleDir: {Level: rfc2119.Must, Reference: containerFormatRef},
 
-	// Config.md
-	// Specification Version
-	SpecVersion: {Level: rfc2119.Must, Reference: specVersionRef},
+	// config.md
+	// Specification version
+	SpecVersionInSemVer: {Level: rfc2119.Must, Reference: specificationVersionRef},
 	// Root
-	RootOnNonHyperV: {Level: rfc2119.Required, Reference: rootRef},
-	RootOnHyperV:    {Level: rfc2119.Must, Reference: rootRef},
-	// TODO: add tests for 'PathFormatOnWindows'
-	PathFormatOnWindows: {Level: rfc2119.Must, Reference: rootRef},
-	PathName:            {Level: rfc2119.Should, Reference: rootRef},
-	PathExistence:       {Level: rfc2119.Must, Reference: rootRef},
-	ReadonlyFilesystem:  {Level: rfc2119.Must, Reference: rootRef},
-	ReadonlyOnWindows:   {Level: rfc2119.Must, Reference: rootRef},
+	RootOnWindowsRequired:      {Level: rfc2119.Required, Reference: rootRef},
+	RootOnHyperVNotSet:         {Level: rfc2119.Must, Reference: rootRef},
+	RootOnNonHyperVRequired:    {Level: rfc2119.Required, Reference: rootRef},
+	RootPathOnWindowsGUID:      {Level: rfc2119.Must, Reference: rootRef},
+	RootPathOnPosixConvention:  {Level: rfc2119.Should, Reference: rootRef},
+	RootPathExist:              {Level: rfc2119.Must, Reference: rootRef},
+	RootReadonlyImplement:      {Level: rfc2119.Must, Reference: rootRef},
+	RootReadonlyOnWindowsFalse: {Level: rfc2119.Must, Reference: rootRef},
+	// Mounts
```

Comment:
The constants now have namespaced names (e.g. `Mounts*`), so there's no need for comments like `// Mounts`, your earlier `// Root`, etc.  There may still be a use for the file-level comments (`// Bundle.md` and `// Config.md`) with the current names, but in order to avoid collisions it may be better to include that level in the constant names too (`ConfigFileExistence` → `BundleConfigFileExistence`, `RootOnWindowsRequired` → `ConfigRootOnWindowsRequired`, etc.).